### PR TITLE
Many fixes for reliability of the scraper

### LINKED ISSUE #74: Do not create a ZIM when failure encountered
See https://farm.openzim.org/pipeline/0810df41-3332-4b81-8687-b67648c1ffcd which failed but produced a ZIM ... bad ...

### LINKED ISSUE #76: Handle case where assets are "bad"
We've had following exception in https://farm.openzim.org/pipeline/dc8c87a5-dfda-4b84-bc35-4fa2032b1d43

```
requests.exceptions.ConnectionError: HTTPConnectionPool(host='mygeologypage.ucdavis.edu', port=80): Max retries exceeded with url: /sumner/gel109/labs/USGSExplanation.jpg (Caused by NameResolutionError("<urllib3.connection.HTTPConnection object at 0x7f5750c031a0>: Failed to resolve 'mygeologypage.ucdavis.edu' ([Errno -2] Name or service not known)"))
```

Looking online, http://mygeologypage.ucdavis.edu/sumner/gel109/labs/USGSExplanation.jpg, the host is really not reachable. 

It might even be permanently down.

How should we handle such things where the asset (an image of one page) is not reachable. We want to avoid creating ZIMs with broken images, but here the image is broken at the source.

I propose to:
- add a flag with the regex of known bad assets URLs 
- add a "bad image here" image in the ZIM (to be designed ...)
- replace all the assets matching the regex with a redirect to the "bad image here" image

### LINKED ISSUE #77: Do not fail when index is empty
See https://farm.openzim.org/pipeline/4049d225-564f-4c93-99ee-eb3ffae8e7f7/debug

Page ID: 15851
URL: https://stats.libretexts.org/Courses/Diablo_Valley_College/Math_142%3A_Elementary_Statistics_(Kwai-Ching)/Math_142%3A_Course_Material/zz%3A_Back_Matter/10%3A_Index

The index is empty online, and the body is empty, this is in fact "normal", we should not fail the scraper but just render an empty index.

### PR DESCRIPTION
**Pull request by benoit74**
Fix #74
Fix #76
Fix #77 (and glossary had the same problem)

Workaround for #71 (real solution postponed to "later") and many other likely situations where we encounter an "unknown" src/href/srcset (inline JS and CSS, ...)

Changes: see list of commits 

Some remarks:
- with these changes, we tend to more often display only warning in the logs for situation which are known to not be totally OK but not cause significant harm to the final ZIM ; this is expected to be able to create a ZIM even when weird situation are encountered in the wild ; I tend to prefer these warnings to really handling the situation to be more easily able to investigate without debugging (e.g. for empty index or glossary pages, I preferred to log a warning and be able to investigate rather than just handling the case where it might be empty and saying nothing, this is not really normal and might be a real problem, we just don't know without manual analysis)
- I dropped the `--html-issues-warn-only` because now we do not have such issues anymore, we are always just logging a warning because these issues are here to stay and we do not want to fail the scrape just for few isolated issues (they are really isolated from test scrape which ran last days)
- as discussed, we know have a new dev option to control how many assets are allowed to fail (to provide some slack when needed) and a prod option to register known bad assets (and we automatically include some known ones coming from automatic CSS)

### CODE REVIEW

File: scraper/src/mindtouch2zim/processor.py [hunk 1]
```diff
@@ -254,184 +255,189 @@ def run(self) -> Path:
...
+                            "Root page is private, we cannot ZIM it, stopping"
+                        ) from None
+                    logger.debug(f"Ignoring page {page.id} (private page)")
+                    private_pages.append(page)
+                    continue
+        logger.info(f"{len(private_pages)} private pages have been ignored")
+        if len(private_pages) == len(selected_pages):
+            # we should never get here since we already check fail early if root
+            # page is private, but we are better safe than sorry
+            raise Exception("All pages have been ignored, not creating an empty ZIM")
```
Comment (rgaudin) [inline scraper/src/mindtouch2zim/processor.py:420]: Use OSError instead

File: scraper/src/mindtouch2zim/asset.py [hunk 2]
```diff
@@ -89,11 +98,28 @@ def _process_asset_internal(
...
+                    self.bad_assets_count += 1
+                    if (
+                        self.bad_assets_threshold >= 0
+                        and self.bad_assets_count > self.bad_assets_threshold
+                    ):
+                        logger.error(
+                            f"Exception while processing asset for {asset_url.value}: "
+                            f"{exc}"
+                        )
+                        raise Exception(  # noqa: B904
```
Comment (rgaudin) [inline scraper/src/mindtouch2zim/asset.py:114]: We should use more qualified exceptions such as `OSError`

**Review by rgaudin — APPROVED** [review]
LGTM ; see suggestion to rename Exception

**Comment by benoit74** [comment]
I've opened https://github.com/openzim/mindtouch/issues/91 for the `Exception`, there are many more than the ones you're mentioning here. But good point indeed

**Pull request merged**